storage/store_pool.go

```diff
@@ -0,0 +1,298 @@
+	n := len(*pq)
+	item := x.(*storeDetail)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+// Pop implements the heap.Interface.
+func (pq *storePoolPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// peek returns the next value in the priority queue without dequeuing it.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) peek() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := (*pq)[0]
+	return detail
+}
+
+// enqueue either adds the detail to the queue or updates its location in the
+// priority queue.
+// Lock is assumed held by caller.
+func (pq *storePoolPQ) enqueue(detail *storeDetail) {
+	if detail.index < 0 {
+		heap.Push(pq, detail)
+	} else {
+		heap.Fix(pq, detail.index)
+	}
+}
+
+// dequeue removes the next detail from the priority queue.
+// Lock is assumed help by the caller.
+func (pq *storePoolPQ) dequeue() *storeDetail {
+	if len(*pq) == 0 {
+		return nil
+	}
+	detail := heap.Pop(pq).(*storeDetail)
+	return detail
+}
+
+// StorePool maintains a list of all known stores in the cluster and
+// information on their health.
+type StorePool struct {
+	gossip             *gossip.Gossip
+	clock              *hlc.Clock
+	timeUntilStoreDead time.Duration
+
+	mu     sync.RWMutex // Protects variables below...
+	stores map[proto.StoreID]*storeDetail
+	queue  storePoolPQ
+}
+
+// NewStorePool creates a StorePool and registers the store updating callback
+// with gossip.
+func NewStorePool(g *gossip.Gossip, timeUntilStoreDead time.Duration, stopper *stop.Stopper) *StorePool {
+	sp := &StorePool{
+		timeUntilStoreDead: timeUntilStoreDead,
+		stores:             make(map[proto.StoreID]*storeDetail),
+	}
+	heap.Init(&sp.queue)
+
+	storeRegex := gossip.MakePrefixPattern(gossip.KeyStorePrefix)
+	g.RegisterCallback(storeRegex, sp.storeGossipUpdate)
+
+	sp.start(stopper)
+
+	return sp
+}
+
+// storeGossipUpdate The gossip callback used to keep the StorePool up to date.
+func (sp *StorePool) storeGossipUpdate(_ string, content []byte) {
+	storeDesc := new(proto.StoreDescriptor)
+	if err := gogoproto.Unmarshal(content, storeDesc); err != nil {
+		log.Error(err)
+		return
+	}
+
+	sp.mu.Lock()
+	defer sp.mu.Unlock()
+	// Does this storeDetail exist yet?
+	if detail, ok := sp.stores[storeDesc.StoreID]; ok {
+		detail.markAlive(time.Now(), *storeDesc)
+		sp.queue.enqueue(detail)
+	} else {
+		newDetail := &storeDetail{
+			desc:            *storeDesc,
+			dead:            false,
+			lastUpdatedTime: time.Now(),
+			index:           -1, // to ensure this gets added to the queue
+		}
+		sp.stores[storeDesc.StoreID] = newDetail
+		sp.queue.enqueue(newDetail)
+	}
+}
+
+// start will run continuously and mark stores as offline if they haven't been
+// heard from in longer than timeUntilStoreDead.
+func (sp *StorePool) start(stopper *stop.Stopper) {
+	stopper.RunWorker(func() {
+		for {
+			var timeout time.Duration
+			sp.mu.Lock()
```

Comment:
I've update the comment on the mutex already.  But it's not just the gossip callback and this. In a multi-store per node store system, the allocator, gossip callback, replica scanner and possibly some health stats will all be using a single store pool.